hyperlink/_url.py

```diff
@@ -1560,20 +1560,26 @@ def get(self, name):
         """
         return [value for (key, value) in self.query if name == key]
 
-    def remove(self, name):
+    def remove(self, name, value=_UNSET):
         """Make a new :class:`URL` instance with all occurrences of the query
-        parameter *name* removed. No exception is raised if the
+        parameter *name* removed, or, if *value* is set, parameters
+        matching *name* and *value*. No exception is raised if the
         parameter is not already set.
 
         Args:
             name (unicode): The name of the query parameter to remove.
+            value (unicode): Optional value to additionally filter
+               on. Setting this removes query parameters which match
+               both name and value.
 
         Returns:
             URL: A new :class:`URL` instance with the parameter removed.
-
         """
-        return self.replace(query=((k, v) for (k, v) in self.query
-                                   if k != name))
+        if value is _UNSET:
+            nq = [(k, v) for (k, v) in self.query if k != name]
```

Comment:
This was previously, extremely intentionally, a generator expression so that the `replace` call wouldn't need to build a temporary list and then replace it with a tuple, it could avoid the intermediate allocation and go straight to the tuple.  Can you make these genexps instead?

Reply:
Did you bench that?

Lists (and list comps) are faster than genexps for low-level operations like this, even if there's an extra intermediate list being constructed:

```
$ python -m timeit -s "x = [1, 2, 3, 4, 5, 6, 7]" "tuple([y for y in x])"
1000000 loops, best of 3: 0.319 usec per loop
$ python -m timeit -s "x = [1, 2, 3, 4, 5, 6, 7]" "tuple(y for y in x)"
1000000 loops, best of 3: 0.631 usec per loop
$ python -m timeit -s "x = list(range(1000))" "tuple([y for y in x])"
100000 loops, best of 3: 19 usec per loop
$ python -m timeit -s "x = list(range(1000))" "tuple(y for y in x)"
10000 loops, best of 3: 36.7 usec per loop
```

(yes on py2 and py3) 

URLs are a pretty clear-cut case of being weighted toward short lists. Most clients have URL length limits and also the limits of clipboards and human readability.

I've tried in the past not to generate extra lines of diff in such a closely-scrutinized library, but I'm not going to unlearn my micro-optimization muscle memory. I'll probably change others when I touch those lines, and request others do the same.

If you have other such cases of extreme intention outside of hyperlink, I would revisit them to see if you're losing cycles. Generators are like a once a month thing, outside of inlineCallbacks and other nice high-level stuff. :v:

Reply:
@glyph my turn to bump. ;) I've resolved Alex's comments, and addressed the above, so I think this is ready to merge. Y/n?